File: Codes in C++/Dynamic Programming/Grid Unique Paths with Obstacle.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
    int m=obstacleGrid.size();
    int n=obstacleGrid[0].size();
    vector<int> prev(n,0);
        for(int i=0;i<m;i++)
        {   vector<int> temp(n,0);
            for(int j=0;j<n;j++)
            {   
                if(obstacleGrid[i][j]==1) 
                    temp[j]=0;
                else if(i==0&&j==0) 
                    temp[0]=1;
                else {
                if(i>0)
                    temp[j]+=prev[j];
                if(j>0)
                    temp[j]+=temp[j-1];
                }  
            }
            prev=temp;
        }
    return prev[n-1];  
    }
};

File: Codes in Python3/Dynamic Programming/Grid Unique Paths with Obstacle.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m=len(obstacleGrid)
        n=len(obstacleGrid[0])
        prev=[0]*(n)
        for i in range(m):
            temp=[0]*(n)
            for j in range(n):
                if obstacleGrid[i][j]==1:
                    temp[j]=0
                elif i==0 and j==0:
                    temp[j]=1
                else:
                    if(i>0):
                        temp[j]+=prev[j]
                    if(j>0):
                        temp[j]+=temp[j-1]
            prev=temp
        return prev[n-1]
        